text/3678-final.md

```diff
@@ -0,0 +1,123 @@
+- Feature Name: `final`
+- Start Date: 2024-07-20
+- RFC PR: [rust-lang/rfcs#3678](https://github.com/rust-lang/rfcs/pull/3678)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Support restricting implementation of individual methods within traits, using
+the already reserved `final` keyword.
+
+# Motivation
+[motivation]: #motivation
+
+When defining a trait, the trait can provide optional methods with default
+implementations, which become available on every implementation of the trait.
+However, the implementer of the trait can still provide their own
+implementation of such a method. In some cases, the trait does not want to
+allow implementations to vary, and instead wants to guarantee that all
+implementations of the trait use an identical method implementation. For
+instance, this may be an assumption required for correctness.
+
+This RFC allows restricting the implementation of trait methods.
+
+This mechanism also faciliates marker-like traits providing no implementable
+methods, such that implementers only choose whether to provide the trait and
+never how to implement it; the trait then provides all the method
+implementations.
+
+One example of a trait in the standard library benefiting from this:
+`Error::type_id`, which has thus far remained unstable because it's unsafe to
+override. This RFC would allow stabilizing that method so users can call it,
+without permitting reimplementation of it.
+
+# Explanation
+[explanation]: #explanation
+
+When defining a trait, the definition can annotate methods or associated
+functions to restrict whether implementations of the trait can define them. For
+instance:
+
+```rust
+trait MyTrait: Display {
+    final fn method(&self) {
+        println!("MyTrait::method: {self}");
+    }
+}
+```
+
+A method or associated function marked as `final` must have a default body.
+
+When implementing a trait, the compiler will emit an error if the
+implementation attempts to define any method or associated function marked as
+`final`, and will emit a suggestion to delete the implementation.
+
+In every other way, a `final` method or associated function acts identically to
+any other method or associated function, and can be invoked accordingly:
+
+```rust
+fn takes_mytrait(m: &impl MyTrait) {
+    m.method();
+}
+```
+
+Note that in some cases, the compiler might choose to avoid placing a `final`
+method in the trait's vtable, if the one-and-only implementation does not
+benefit from monomorphization.
+
+Note that removing a `final` restriction is always forwards-compatible.
+
+The keyword `final` has been reserved since Rust 1.0, so this feature can ship
+identically in all editions.
+
```

Comment:
It should also state that `final` has no impact on dyn compatibility of the trait.

At first I thought it would, but I realized that it shouldn't.. if you have a self-by-value method that's not going to work with `dyn Trait` as the self type regardless of whether the method is final. So you still need `where Self: Sized` (and in this particular case, a bridging `impl Trait for Box<dyn Trait>` automatically gets you back the method).

Reply:
@scottmcm I tend to remove that section (and retitle the remaining one as "Explanation") when there isn't any meaningful difference in content for it. But the items that you're suggesting seem like reasonable additions to me. Thanks! I'll merge and then do some edits.